org/mozilla/jss/netscape/security/util/Cert.java

```diff
@@ -204,4 +214,130 @@ public static String normalizeCertStrAndReq(String s) {
         String b64 = stripBrackets(encoded);
         return Utils.base64decode(b64);
     }
+
+    /**
+     * Sorts certificate chain from root to leaf.
+     *
+     * This method sorts an array of certificates (e.g. from a PKCS #7
+     * data) that represents a certificate chain from root to leaf
+     * according to the subject DNs and issuer DNs.
+     *
+     * The input array is a set of certificates that are part of a
+     * chain but not in specific order.
+     *
+     * The result is a new array that contains the certificate chain
+     * sorted from root to leaf. The input array is unchanged.
+     *
+     * @param certs input array of certificates
+     * @return new array containing sorted certificates
+     */
+    public static java.security.cert.X509Certificate[] sortCertificateChain(java.security.cert.X509Certificate[] certs) throws Exception {
+
+        // lookup map: subject DN -> cert
+        Map<String, java.security.cert.X509Certificate> certMap = new LinkedHashMap<>();
+
+        // hierarchy map: subject DN -> issuer DN
+        Map<String, String> parentMap = new HashMap<>();
+
+        // reverse hierarchy map: issuer DN -> subject DN
+        Map<String, String> childMap = new HashMap<>();
+
+        // build maps
+        for (java.security.cert.X509Certificate cert : certs) {
+
+            String subjectDN = cert.getSubjectDN().toString();
+            String issuerDN = cert.getIssuerDN().toString();
+
+            if (certMap.containsKey(subjectDN)) {
+                throw new Exception("Duplicate certificate: " + subjectDN);
+            }
+
+            certMap.put(subjectDN, cert);
+
+            // ignore self-signed certificate
+            if (subjectDN.equals(issuerDN)) continue;
+
+            if (childMap.containsKey(issuerDN)) {
+                throw new Exception("Branched chain: " + issuerDN);
+            }
+
+            parentMap.put(subjectDN, issuerDN);
+            childMap.put(issuerDN, subjectDN);
+        }
+
+        if (logger.isDebugEnabled()) {
+            logger.debug("Certificates:");
+            for (String subjectDN : certMap.keySet()) {
+                logger.debug(" - " + subjectDN);
+
+                String parent = parentMap.get(subjectDN);
+                if (parent != null) logger.debug("   parent: " + parent);
+
+                String child = childMap.get(subjectDN);
+                if (child != null) logger.debug("   child: " + child);
+            }
+        }
+
+        // find leaf cert
+        List<String> leafCerts = new ArrayList<>();
+
+        for (String subjectDN : certMap.keySet()) {
+
+            // if cert has a child, skip
+            if (childMap.containsKey(subjectDN)) continue;
+
+            // found leaf cert
+            leafCerts.add(subjectDN);
+        }
+
+        if (leafCerts.isEmpty()) {
+            throw new Exception("Unable to find leaf certificate");
+        }
+
+        if (leafCerts.size() > 1) {
```

Comment:
@edewata One additional thought while I'm here, regarding future use of JSS for a `PKICertImport` replacement.

OpenSSL only creates `.p12` files with one chain and one key in them. Java and a few other providers lets you stick multiple chains/keys in one `.p12` file.

Do we want to extend this to support multiple chains, and leave chain quantity validation as a separate function? Or leave that to a later time?

Reply:
I guess the behavior is non-trivial:

Do we want a list of complete chains? Or just a topological sort of the graph? Which we want probably depends on usage... But guessing topological sort of the graph would be best.

Reply:
Our tools can export the whole NSS database (which may contain multiple chains/keys) into a PKCS #12 file, which can be used to set up a clone or for migration. If we need to create a topological graph we can certainly refactor the code, but we still need to keep this method to sort a single chain from PKCS #7 data. It should be easy to do that separately later. This patch is mainly for moving the code, so I'd like to minimize code changes in this patch.

Reply:
Fine by me. Thanks!